### PR TITLE
feat: inventory module

### LINKED ISSUE #217: Interface Mappings extension | Device-Type mapping
### NetBox Librenms Plugin version

0.4.2

### NetBox version

v4.5.2

### Feature type

Change to existing functionality

### Proposed functionality

It would be nice to have more (optional) filters to match when doing interface mapping - there are some cases when it would be good to include Device-Type and/or Module-Type - because that determines actual interface type together with speed/Librenms type.
** edit ** - actually, just taking netbox type into account would be probably best solution. 

I have similar issue with device-type, for example in LibreNMS: Juniper MX480 Internet Backbone Router
but in Netbox I'd have MX480, maybe part_number CHAS-BP-MX480-S  - but not what LibreNMS has.

### Use case

Image 
but this can be quite a bit of different 10G types. 
On interface - there's a Type field - so maybe that could be used as a 'hint' what should it be if populated.

For Device-Type - LibreNMS inventory would be a bit better - but still not a guarantee - so maybe having a mapping would make sense.

Cheers,
Marcin

### External dependencies

_No response_

### PR DESCRIPTION
**Pull request by marcinpsk**
## Summary
Adds a Modules / Inventory sync tab to Device and VM detail pages.  LibreNMS ENTITY-MIB inventory data is fetched, matched against NetBox module types and bays via configurable mapping rules, and presented in a table where users can install, replace, or move modules directly.

**Mapping models (migration 0010)**
- `DeviceTypeMapping` — maps LibreNMS hardware strings → NetBox DeviceTypes
- `ModuleTypeMapping` — maps LibreNMS model strings → NetBox ModuleTypes
- `ModuleBayMapping` — maps LibreNMS inventory names/positions → NetBox ModuleBays
- `NormalizationRule` — rewrites raw LibreNMS model names before lookup (regex or exact, scoped by manufacturer)
- `InventoryIgnoreRule` — suppresses inventory rows that should never sync (e.g. IDPROM, power supplies)
- `PlatformMapping` — maps LibreNMS OS strings → NetBox Platforms (checked before exact-name fallback)

**Module sync actions**
- Install a new module into an empty bay
- Replace an existing module (serial/type mismatch)
- Move a module from another bay when a serial conflict is detected (with ambiguity detection for duplicate serials)

**Mapping management UI**
- Full CRUD views for all six mapping models
- YAML bulk export for sharing/versioning mapping sets
- Navigation menu entries under a new Mappings group

**LibreNMS API**
- `get_device_transceivers()` endpoint for transceiver inventory data

**contrib/**
- Example YAML files for device type, module type, module bay, interface type mappings, normalization rules, and inventory ignore rules

**Tests**
- new test code covering sync logic, mapping CRUD, module actions, table rendering, API serializers, views, filters, forms, caching, and platform matching

Would close #217
Depends on #259 #258 

## Motivation / Problem
- Feature

Closes the gap between LibreNMS hardware inventory and NetBox module tracking. Previously users had to manually create and assign modules; this automates discovery and placement using configurable mapping rules.

## Scope of Change

- Sync/Import logic
- NetBox models / ORM
- LibreNMS API interaction
- Web UI / templates
- Database migrations
- Tests

## How Was This Tested?

- Unit tests: yes — comprehensive pytest suite covering all new models, views, sync logic, mapping resolution, table rendering, API endpoints, and edge cases
- Manual testing: yes — module install, replace, and move actions verified against live LibreNMS instances
- E2E tests: yes — `tests/e2e/test_module_install.py` for install workflow

## Risk Assessment
- Does this change affect existing users?
  No — the feature is additive. Existing sync tabs and import workflows are unchanged. The `librenms_id` migration (0009) converts the custom field format but preserves all existing values.
- Could this cause unintended imports / updates?
  No — module actions (install/replace/move) require explicit user confirmation per row.

## Backwards Compatibility
- No breaking changes

## Other Notes
- The `contrib/` YAML files are examples only and are not auto-loaded — users import them via the bulk export/import UI.
- Mapping resolution order: exact match → normalization rules → generic fallback. Ambiguous matches (multiple candidates) surface a clear status rather than silently picking one.
- Currently there is one table that is auto-loaded - InventoryIgnoreRule - populates Cisco entries that should be ignored - that could be removed.

### CODE REVIEW

**Comment by bonzo81** [comment]
@marcinpsk apologies, I didn't mean to close this. Turned on the auto close branches after PR merge setting in Github, but develop wasn't protected properly.

**Comment by bonzo81** [comment]
Hi @marcinpsk thanks for this major new feature PR 👏.

I've done a preliminary UI test and review. Here's some comments:

### Feature
1. **Virtual chassis aware**. We will need the module table to show virtual chassis member dropdown with some view logic to pre determine which chassis member the module is linked too. This is similar to the interface and cable tables already but will likely need it's own logic to attempt to match virtual chassis member to module details from LibreNMS api data.

### Bugs
1. **The 'Install Selected' button causes a crash** when either a single or multiple row selected. 
```
Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/psycopg/cursor.py", line 117, in execute
    raise ex.with_traceback(None)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The above exception (FOR UPDATE cannot be applied to the nullable side of an outer join) was the direct cause of the following exception:
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/views/generic/base.py", line 105, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/contrib/auth/mixins.py", line 109, in dispatch
    return super().dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/views/generic/base.py", line 144, in dispatch
    return handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/netbox-librenms-plugin/netbox_librenms_plugin/views/sync/modules.py", line 573, in post
    result = InstallBranchView._install_single(
             
  File "/workspaces/netbox-librenms-plugin/netbox_librenms_plugin/views/sync/modules.py", line 326, in _install_single
    ModuleBay.objects.select_for_update().select_related("installed_module").get(pk=matched_bay.pk)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/models/query.py", line 631, in get
    num = len(clone)
          ^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/models/query.py", line 368, in __len__
    self._fetch_all()
    ^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/models/query.py", line 1954, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/models/query.py", line 93, in __iter__
    results = compiler.execute_sql(
              
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 1623, in execute_sql
    cursor.execute(sql, params)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/backends/utils.py", line 122, in execute
    return super().execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
           
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/backends/utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.12/site-packages/psycopg/cursor.py", line 117, in execute
    raise ex.with_traceback(None)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Exception Type: NotSupportedError at /plugins/librenms_plugin/devices/824/install-selected/
Exception Value: FOR UPDATE cannot be applied to the nullable side of an outer join
```

2. **The 'Install Selected' button disappears** after syncing a module row with the 'install' button.
3. When there is a Module Mismatch,  when clicking the 'Replace' button, a modal with text `htmx is not defined` dispalys.

### UI/UX
1. I think we need to **organise the navigation** to group the options logically.  I suggest all mapping items have there own 'Mappings' header. 
2. **Row highlighting** for installed modules look good in light theme but in dark mode it makes the linked text hard to read. Consider removing highlighting the row. The 'installed' badge still clearly shows module state.
3. **Badge color classes** need to be theme aware also. Suggest adding text contrast classes; e.g.
```  
badge_classes = {
            "Installed": "bg-success text-white",
            "Matched": "bg-info text-white",
            "No Bay": "bg-warning text-dark",
            "No Type": "bg-warning text-dark",
            "Unmatched": "bg-secondary text-white",
            "No Match": "bg-secondary text-white",
            "Serial Mismatch": "bg-danger text-white",
            "Name Conflict": "bg-warning text-dark",
            "Type Mismatch": "bg-warning text-dark",
        }
```

**Comment by marcinpsk** [comment]
hi @bonzo81 

Thanks for having a look! 

Bugs fixed:                                                                                                                                                
  - "Install Selected" crash — FOR UPDATE on outer join, fixed with select_for_update(of=("self",))
  - "Install Selected" button disappearing after a per-row install — has_write_permission was missing from the full-page context
  - Replace modal showing "htmx is not defined" — unguarded htmx.process() call                                                 
                                                                                                                                                             
  Dark mode / readability:                                                                                                                                   
  - Removed all row background highlighting from the module table — badges carry the state - looks much cleaner
  - Added text-white/text-dark contrast classes to all status badges
                                                                                                                                                             
  Navigation:                                                                                                                                                
  - Moved all mapping types into a dedicated "Mappings" group                                                                                                
  - Reordered groups: Import → Status Check → Mappings → Settings  -- that's just a proposal - looked better for me - but maybe you prefer another order


As for VC - I'm open for ideas, but I have no VC so it's making any testing difficult.

File: netbox_librenms_plugin/views/imports/actions.py [hunk 1]
```diff
@@ -1392,6 +1400,154 @@
...
+            # A <tr> following a <div> is invalid HTML and gets silently dropped by the browser
+            # parser when HTMX wraps the combined response in a <template> for parsing. Wrapping
+            # in <table><tbody> keeps the <tr> in a valid table context so HTMX finds and applies
+            # the OOB swap. The <div id="django-messages"> inside is foster-parented outside the
+            # table by the parser, so both OOB elements are preserved.
+            row_html = f"<table><tbody>{row_html}</tbody></table>"
+        else:
+            row_html = ""
+
+        return HttpResponse(oob_modal + row_html, content_type="text/html")
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_librenms_plugin/views/imports/actions.py:1765]: Valid finding — the URL converter was `<str:device_id>` which made `device_id` a user-controlled string. It was embedded in the HTML response via the `row_html.replace()` f-string and the HTMX OOB fragment, giving CodeQL a clear taint path to the `HttpResponse` sink.

Fixed in `188b0bc` : changed all seven `device-import/*` URL patterns from `<str:device_id>` to `<int:device_id>`. LibreNMS device IDs are always integers; Django's URL router now rejects non-integer values at the routing level, removing all four taint paths. Templates use `libre_device.device_id` (integer from LibreNMS API) so no template changes were needed.

**Comment by bonzo81** [comment]
Hi @marcinpsk, thanks for continued updates on this. I've been working on adding virtual chassis logic to this PR. Got something working, so I'll commit the the PR shortly.

File: netbox_librenms_plugin/views/imports/actions.py [hunk 2]
```diff
@@ -1392,6 +1400,158 @@
...
+            # table by the parser, so both OOB elements are preserved.
+            row_html = f"<table><tbody>{row_html}</tbody></table>"
+        else:
+            row_html = ""
+
+        # Both oob_modal and row_html are produced by Django template rendering,
+        # which auto-escapes all user-supplied values. CodeQL cannot model
+        # Django's template-engine escaping as a sanitizer and flags the
+        # request → template-render → response chain as reflected XSS.
+        return HttpResponse(oob_modal + row_html, content_type="text/html")  # lgtm[py/reflected-xss]
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_librenms_plugin/views/imports/actions.py]: CodeQL is now **green** on PR #261 (all checks pass at `96a89e1`). The alert was fixed in `6ba8308` by using `format_html()` + `mark_safe()` to compose the OOB response — CodeQL's Django taint model recognises this pattern and stops tracking the taint through template-rendered output.

**Comment by marcinpsk** [comment]
hi @bonzo81 - great! maybe one day I'll find some time to get some VMs where  I could simulate virtual chassis - but it's nice you got something!

**Comment by bonzo81** [comment]
@marcinpsk pushed the changes to make the module table virtual chassis aware. Feel free to do any quality reviews you want to.

Seperately, I am wondering how this feature handles interface modules like SFPs. For example:
image

Do we need to add interface identification also? The Librenms port name looks like the ifName value, so might not be an exact match for interfaces imported using ifDescr. What's you thoughts on handling this scenario?

**Comment by marcinpsk** [comment]
@bonzo81 I'm not sure how to handle this. Each device is just different. This comes from different SNMP MIB - and for example I have:
image

It is mostly Cisco/Nokia I've seen where those do differ. 
But not all.
image
some Cisco (I saw on IOS-XR) - same for both - ifDescr and ifName is the same.

What I was thinking of adding later - persistent per-device selection of ifDesc/ifName - so it would persist it per device at least in UI - some small QoL to help with getting at least consistent names. 

But - what interface you create is determined by model, right? it's the same if you would insert module into device manually - it would create what you defined on a model. This is just to get a match into correct slot.

for myself for more complex scenarios - I've made plugin to rename interfaces based on some criteria - mostly to handle channelized interfaces - or ones that have strange numbering rules (like with offset etc) - and to do it whichever way it is added (librenms/manually/API). 


But how to actually get sync of your model and interfaces you're getting... didn't think about it. Modules would create/update interfaces and interface path would create/update interface. If model matches snmp interface information - that will be the same one. 
But if this is not matching... I guess changing Netbox model is right approach. And being consistent with which field is being used when importing interfaces.

Could maybe check on interface import to do some soft check if interface that would get created would match model, idk.

Do you have any thoughts about it?

**Comment by marcinpsk** [comment]
and about SFP - it's just a module with usually one interface - same way - if you have matching bay (and most of the time you do have to define module bay mapping per HW model - that's why I added some examples in contrib/ - but maybe I can add a few more for SFPs - I think I added only 1 or 2)

**Comment by marcinpsk** [comment]
hi @bonzo81 
I've fixed some of issues I've noticed with some of vendors I have (for example it's fetching transceiver information not from librenms inventory, but from transceivers because they're not exposed in inventory - and it wasn't matching it correctly).

Added some quality of life improvements as well - as being able to add a mapping directly from row + some predictions how that mapping should look like based on what's in row.
Plus some checks around device type model - to make it easier to correct model.
No changes around VC.

**Comment by marcinpsk** [comment]
@bonzo81 - one thing I was thinking of still changing - squashing all migrations introduced in this PR into one file. This may have some issues in devcontainers that already have some migrations done.
You may need:
python manage.py migrate netbox_librenms_plugin 0009 --fake
python manage.py migrate netbox_librenms_plugin --fake

**Comment by bonzo81** [comment]
@marcinpsk Thanks for the update. Good idea re the migration squash. Thanks for the devcontainer heads up.

I've been working on handling the interface matching alongside SFPs and/or module type that potentially have interface templates. Using the port_id/librenms_id as the anchor for primary identification. It's a beast! But getting something working now and will push here for further testing and refinement soon. I'll pull you latest changes first.

This is a big feature set but it will be a great addition once merged. We will need to ensure the docs are in good order for the this feature to cover different scenarios/device types/module types, but that could be a separate PR.

**Comment by marcinpsk** [comment]
@bonzo81 I've changed quite a bit around module bay matching etc - not really around interfaces exactly - but that can have impact on this.

**Comment by bonzo81** [comment]
@marcinpsk The migration dependency ` 0227_alter_interface_speed_bigint`  will raise the minimum Netbox version for the plugin to NetBox v4.5.7 from what I can see on Netbox releases. Just identifying that for awareness on our next plugin release.

**Comment by marcinpsk** [comment]
@bonzo81 it was auto picked by makemigrations - we could probably drop it down - to dcim.0200_populate_mac_addresses + extras.0122_charfield_null_choices
- would have to check

**Comment by marcinpsk** [comment]
@bonzo81 I've dropped it down to match 4.2 - I don't see anything that would need any newer.

**Comment by bonzo81** [comment]
Hi @marcinpskk, I'm going to move my module interface creation changes to a new PR so we can merge this one. That will allow for the other PRs coming after it to progress. I'll kick off Copilot review now as a review final pass.

**Comment by marcinpsk** [comment]
@bonzo81 unfortunately, it's too big for copilot - I'm using for some of mine Coderabbit - it has a free plan for Open Source projects - with hourly limits, but can review 50k as far as I remember - and it's always like 2nd pair of eyes to Copilot.

**Comment by bonzo81** [comment]
Thanks @marcinpsk I'll take a look at CodeRabbit. sound good

**Comment by bonzo81** [comment]
@coderabbitai full review

**Comment by marcinpsk** [comment]
@bonzo81 looks clean.

**Comment by bonzo81** [comment]
Hi @marcinpsk, thanks for all the work on this. Lets merge it. Work busy just now, but when I get time I will add a PR for interface handling to the module sync page

**Comment by marcinpsk** [comment]
@bonzo81 sounds good!
